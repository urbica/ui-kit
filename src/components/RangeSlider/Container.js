import styled from 'react-emotion';


export default styled.div`
  position: relative;
  height: 34px;

  input {
    display: block;
    position: absolute;
    top: 0;
    left: 0;
  }
`;

