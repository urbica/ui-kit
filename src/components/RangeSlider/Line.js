import styled from 'react-emotion';


export default styled.div`
  position: absolute;
  height: 2px;
  top: calc(50% - 1px);
  left: ${props => props.percentLeft || 0}%;
  right: ${props => props.percentRight || 0}%;
  background: black;
`;

