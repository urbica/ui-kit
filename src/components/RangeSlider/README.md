RangeSlider

```js
initialState = { leftValue: 20, rightValue: 100 };
const onChange = (value) => setState(value);

<div>
  <RangeSlider
    onChange={onChange}
    leftValue={state.leftValue}
    rightValue={state.rightValue}
  />
  Current value: {state.leftValue}, {state.rightValue}
</div>
```

RangeSlider + fixLeft

```js
initialState = { value: 60 };
const onChange = (value) => setState({ value: value.rightValue });

<div>
  <RangeSlider
    onChange={onChange}
    rightValue={state.value}
    fixedLeft
  />
  Current value: {state.value}
</div>
```

RangeSlider + fixRight

```js
initialState = { value: 40 };
const onChange = (value) => setState({ value: value.leftValue });

<div>
  <RangeSlider
    onChange={onChange}
    leftValue={state.value}
    fixedRight
  />
  Current value: {state.value}
</div>
```

RangeSlider (options)

```js
const options = [
  { value: 'one', label: 'One' },
  { value: 'two', label: 'Two' },
  { value: 'three', label: 'Three' },
  { value: 'four', label: 'Four' }
];

initialState = {
  leftValue: options[0].value,
  rightValue: options[2].value
};

const onChange = (value) => setState(value);

<div>
  <RangeSlider
    onChange={onChange}
    leftValue={state.leftValue}
    rightValue={state.rightValue}
    options={options}
  />
  Current value: {state.leftValue}, {state.rightValue}
</div>
```