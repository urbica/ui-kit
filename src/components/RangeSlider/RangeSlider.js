import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import InputRange from '../InputRange/InputRange';
import Container from './Container';
import Line from './Line';
import findIndex from '../../utils/findIndex';

class RangeSlider extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      leftValue: this.props.leftValue,
      rightValue: this.props.rightValue
    };

    this.onChangeLeft = this.onChangeLeft.bind(this);
    this.onChangeRight = this.onChangeRight.bind(this);
    this.onChangeEnd = this.onChangeEnd.bind(this);
  }

  componentWillMount() {
    const { options, leftValue, rightValue } = this.props;

    if (options) {
      if (!options.length) {
        throw new Error('options is empty');
      }

      const leftIndex = leftValue ?
        findIndex(options, o => o.value === leftValue) : 0;
      const rightIndex = rightValue ?
        findIndex(options, o => o.value === rightValue) : options.length - 1;

      this.setState({
        leftValue: leftIndex,
        rightValue: rightIndex
      });
    }
  }

  onChangeLeft(event) {
    const { options, onChange, fixedRight } = this.props;
    const { rightValue } = this.state;
    const tmp = parseFloat(event.target.value);
    const roundTmp = Math.round(tmp);

    if (tmp < rightValue || fixedRight) {
      this.setState({ leftValue: tmp });
      onChange({
        leftValue: options ? options[roundTmp].value : roundTmp,
        rightValue: options ? options[rightValue].value : rightValue
      });
    }
  }

  onChangeRight(event) {
    const { options, onChange, fixedLeft } = this.props;
    const { leftValue } = this.state;
    const tmp = parseFloat(event.target.value);
    const roundTmp = Math.round(tmp);

    if (tmp > leftValue || fixedLeft) {
      this.setState({ rightValue: tmp });
      onChange({
        leftValue: options ? options[leftValue].value : leftValue,
        rightValue: options ? options[roundTmp].value : roundTmp
      });
    }
  }

  onChangeEnd() {
    const { options } = this.props;
    const { leftValue, rightValue } = this.state;
    const roundLeft = Math.round(leftValue);
    const roundRight = Math.round(rightValue);

    this.setState({
      leftValue: roundLeft,
      rightValue: roundRight
    });

    this.props.onChange({
      leftValue: options ? options[roundLeft].value : roundLeft,
      rightValue: options ? options[roundRight].value : roundRight
    });
  }

  render() {
    const { leftValue, rightValue } = this.state;
    const {
      step,
      fixedLeft,
      fixedRight,
      options
    } = this.props;

    const left = options ? 0 : this.props.left;
    const right = options ? options.length - 1 : this.props.right;
    const range = right - left;
    const percentLeft = ((leftValue - left) / range) * 100;
    const percentRight = ((right - rightValue) / range) * 100;

    return (
      <Container>
        <InputRange
          onChange={this.onChangeLeft}
          value={fixedLeft ? 0 : leftValue}
          onClick={this.onChangeEnd}
          onMouseLeave={this.onChangeEnd}
          onTouchEnd={this.onChangeEnd}
          min={left}
          max={right}
          step={step || 0.01}
          type="range"
          fix={fixedLeft}
        />
        <InputRange
          onChange={this.onChangeRight}
          value={fixedRight ? 0 : rightValue}
          onClick={this.onChangeEnd}
          onMouseLeave={this.onChangeEnd}
          onTouchEnd={this.onChangeEnd}
          min={left}
          max={right}
          step={step || 0.01}
          type="range"
          fix={fixedRight}
        />
        <Line
          percentLeft={fixedLeft ? 0 : percentLeft}
          percentRight={fixedRight ? 0 : percentRight}
        />
      </Container>
    );
  }
}

RangeSlider.propTypes = {
  leftValue: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.shape({
      value: PropTypes.node.isRequired,
      label: PropTypes.node
    })
  ]),
  rightValue: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.shape({
      value: PropTypes.node.isRequired,
      label: PropTypes.node
    })
  ]),
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.node.isRequired,
    label: PropTypes.node
  })),
  onChange: PropTypes.func,
  step: PropTypes.number,
  left: PropTypes.number,
  right: PropTypes.number,
  fixedLeft: PropTypes.bool,
  fixedRight: PropTypes.bool
};

RangeSlider.defaultProps = {
  leftValue: 0,
  rightValue: 100,
  options: null,
  step: null,
  left: 0,
  right: 100,
  fixedLeft: false,
  fixedRight: false,
  onChange: () => {}
};

export default RangeSlider;
