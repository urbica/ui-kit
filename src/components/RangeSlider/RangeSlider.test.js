import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import RangeSlider from './';
import Input from '../InputRange/InputRange';

const onChange = value => value;
const leftValue = 0;
const rightValue = 100;

const options = [
  { value: 'one', label: 'One' },
  { value: 'two', label: 'Two' },
  { value: 'three', label: 'Three' },
  { value: 'four', label: 'Four' }
];


test('RangeSlider does not crash', () => {
  const wrapper = shallow(<RangeSlider />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test('RangeSlider with an empty array', () => {
  expect(() => shallow(<RangeSlider onChange={onChange} options={[]} />))
    .toThrowError('options is empty');
});

test('RangeSlider line test', () => {
  const wrapper = mount(<RangeSlider
    options={options}
    leftValue={20}
    rightValue={80}
  />);
  expect(wrapper.props().leftValue).toEqual(20);
  wrapper.setProps({ leftValue: 40 });
  expect(wrapper.props().leftValue).toEqual(40);
});

test('RangeSlider render witch options', () => {
  const wrapper = shallow(<RangeSlider
    options={options}
    leftValue={options[1]}
    rightValue={options[2]}
  />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test('RangeSlider simulate onChange', () => {
  const wrapper = shallow(<RangeSlider
    onChange={onChange}
    leftValue={leftValue}
    rightValue={rightValue}
  />);
  const rangeInputFirst = wrapper.find(Input).first();
  const rangeInputLast = wrapper.find(Input).last();
  expect(wrapper.state('leftValue')).toEqual(0);
  rangeInputFirst.simulate('change', { target: { value: 10 } });
  expect(wrapper.state('leftValue')).toEqual(10);
  rangeInputLast.simulate('change', { target: { value: 80 } });
  expect(wrapper.state('rightValue')).toEqual(80);
});

test('RangeSlider simulate onChangeEnd', () => {
  const wrapper = shallow(<RangeSlider
    onChange={onChange}
    leftValue={leftValue}
    rightValue={rightValue}
  />);
  const rangeInput = wrapper.find(Input).first();
  expect(wrapper.state('leftValue')).toEqual(0);
  rangeInput.simulate('change', { target: { value: 10.8 } });
  rangeInput.simulate('click', { target: { value: 10.8 } });
  expect(wrapper.state('leftValue')).toEqual(11);
});
