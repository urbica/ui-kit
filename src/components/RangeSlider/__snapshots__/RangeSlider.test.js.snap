// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider does not crash 1`] = `
<Styled(div)>
  <Component
    fix={false}
    max={100}
    min={0}
    onChange={[Function]}
    onClick={[Function]}
    onMouseLeave={[Function]}
    onTouchEnd={[Function]}
    step={0.01}
    type="range"
    value={0}
  />
  <Component
    fix={false}
    max={100}
    min={0}
    onChange={[Function]}
    onClick={[Function]}
    onMouseLeave={[Function]}
    onTouchEnd={[Function]}
    step={0.01}
    type="range"
    value={100}
  />
  <Styled(div)
    percentLeft={0}
    percentRight={0}
  />
</Styled(div)>
`;

exports[`RangeSlider render witch options 1`] = `
<Styled(div)>
  <Component
    fix={false}
    max={3}
    min={0}
    onChange={[Function]}
    onClick={[Function]}
    onMouseLeave={[Function]}
    onTouchEnd={[Function]}
    step={0.01}
    type="range"
    value={-1}
  />
  <Component
    fix={false}
    max={3}
    min={0}
    onChange={[Function]}
    onClick={[Function]}
    onMouseLeave={[Function]}
    onTouchEnd={[Function]}
    step={0.01}
    type="range"
    value={-1}
  />
  <Styled(div)
    percentLeft={-33.33333333333333}
    percentRight={133.33333333333331}
  />
</Styled(div)>
`;
